habitat_code/nobodyliveshere-kmeans.py

```diff
@@ -0,0 +1,804 @@
+                    if band_match and band_match.group(1) in desired_bands:
+                        band_urls.append(url_info['URL'])
+
+        if not fmask_url:
+            print(f"No Fmask URL found for granule: {granule_id}")
+            continue  # Skip this granule if no Fmask URL is found
+
+        for url in band_urls:
+            tile_id = re.search(r'T(\d{2}[A-Z]{3})', url).group(1) if re.search(r'T(\d{2}[A-Z]{3})', url) else None
+            band_number = int(re.search(r'B(\d{2})\.tif', url).group(1)) if re.search(r'B(\d{2})\.tif', url) else None
+
+            band_info = {
+                'granule_id': granule_id,
+                'datetime': datetime,
+                'geometry': polygon,
+                'tile_id': tile_id,
+                'band_number': band_number,
+                'band_url': url,
+                'fmask_url': fmask_url
+            }
+            granules_info.append(band_info)
+
+    return gpd.GeoDataFrame(granules_info, geometry='geometry')
+
+
+
+
+# %%
+print(results[0]['umm'])
+
+# %%
+granule_gdf = compile_granule_info(results)
+print(granule_gdf.columns)  # Access columns of the DataFrame
+
+# Grouping by 'granule_id'
+
+if 'processed_data' not in granule_gdf.columns:
+    granule_gdf['processed_data'] = pd.Series(dtype=object)
+
+# Group the DataFrame by 'granule_id'
+grouped_granule_gdf = granule_gdf.groupby('granule_id')
+
+
+for name, group in grouped_granule_gdf:
+    print(f"Columns in group {name}: {group.columns}")
+    break  # Remove this break if you want to see columns for each group
+
+
+# %%
+def process_band_test(band_path, fmask_path, mask_bits, scale_factor, watershed):
+    # Load the Fmask data
+    test_fmask_da = rxr.open_rasterio(fmask_path, masked=True)
+
+    # Ensure the CRS of the watershed matches the CRS of the raster data
+    if watershed.crs != test_fmask_da.rio.crs:
+        watershed = watershed.to_crs(test_fmask_da.rio.crs)
+
+    # Clip the Fmask data to the watershed geometry
+    test_fmask_clipped = test_fmask_da.rio.clip(watershed.geometry)
+
+    # Compute the mask
+    test_bits = np.unpackbits(test_fmask_clipped.astype(np.uint8)).reshape(test_fmask_clipped.shape + (-1,))
+    Test_mask = np.prod(test_bits[..., mask_bits] == 0, axis=-1)
+
+    # Load the band data and clip it to the watershed geometry
+    test_band_da = rxr.open_rasterio(band_path) * scale_factor
+
+    test_band_clipped = test_band_da.rio.clip(watershed.geometry)
+
+
+    # Apply the mask
+    test_masked_band_da = test_band_clipped.where(test_fmask_clipped, np.nan)
+
+    # Check if the mask was applied correctly
+    print(f"Masked band data stats - Min: {test_masked_band_da.min().item()}, Max: {test_masked_band_da.max().item()}, Mean: {test_masked_band_da.mean().item()}, NaNs: {np.isnan(test_masked_band_da).sum().item()}")
+
+    return test_masked_band_da
+
+
+# %%
+# Assuming 'band_url' is the column name where the band paths are stored
+mask_bits = [1, 2, 3, 6, 7]

```

Comment:
It might be helpful for the reader to include here in the code more explicitly why the particular bits are being masked out.